File: Demo(C#)-标准/DC.Demo/SMDemo.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Text;
using System.Web;

namespace DC.Demo
{
    /// <summary>
    ///  国密免前置/SaaS对接示例，本示例仅供参考，不保证各种异常场景运行，请勿直接使用，如有错漏请联系对接人员。运行时，请使用所获取的测试资源替换 用户编号、公私钥、对称密钥、服务商编号等信息。
    /// </summary>
    class SMDemo
    {

        private static string URL = "http://cdctest.cmburl.cn:80/cdcserver/api/v2"; // 银行服务地址(测试)
        private static string bankpubkey = "BNsIe9U0x8IeSe4h/dxUzVEz9pie0hDSfMRINRXc7s1UIXfkExnYECF4QqJ2SnHxLv3z/99gsfDQrQ6dzN5lZj0="; // 银行公钥

        private static string privkey = "NBtl7WnuUtA2v5FaebEkU0/Jj1IodLGT6lQqwkzmd2E=";
        private static string sm4key = "VuAzSWQhsoNqzn0K";//"1234567890123456"; // 用户的对称密钥

        private static string ALG_SM = "SM"; // 采用国密算法

        public static string UID = "N003261207"; // 测试的用户编号

        static void Main(string[] args)
        {

            // 组织发送报文
            JObject obj = new JObject();
            JObject req = new JObject();
            JObject body = new JObject();
            JObject head = new JObject();
            head.Add("funcode", "DCLISMOD");
            head.Add("userid", UID);
            head.Add("reqid", DCHelper.GetTime() + "0000001");
            body.Add("buscod", "N02030");
            body.Add("TEST", "中文");
            body.Add("TEST2", "!@#$%^&*()\\\\///");
            body.Add("TEST3", 12345);
            JArray array = new JArray();
            JObject item = new JObject();
            item.Add("arrItem1", "qaz");
            item.Add("arrItem2", 123);
            item.Add("arrItem3", true);
            item.Add("arrItem4", "中文");

            array.Add(item);
            body.Add("TEST4", array);
            req.Add("head", head);
            req.Add("body", body);
            obj.Add("request", req);

            // 请求发送接收
            doProcess(obj);
        }


        private static void doProcess(JObject jObject)
        {
            JObject obj = new JObject();
            // 签名
            obj.Add("sigdat", "__signature_sigdat__");
            obj.Add("sigtim", DCHelper.GetTime());
            jObject.Add("signature", obj);
            string source = DCHelper.SerialJsonOrdered(jObject);
            Console.WriteLine("签名原文: " + source);
            Encoding encoding = Encoding.UTF8;
            byte[] signature1 = DCCryptor.CMBSM2SignWithSM3(GetID_IV(), Convert.FromBase64String(privkey), encoding.GetBytes(source));
            string sigdat1 = Convert.ToBase64String(signature1);
            Console.WriteLine("签名结果: " + sigdat1);
            obj["sigdat"] = sigdat1;

            // SM4-CBC加密
            string plaintxt = jObject.ToString();
            Console.WriteLine("加密前req:  " + plaintxt);
            byte[] enInput = DCCryptor.CMBSM4EncryptWithCBC(encoding.GetBytes(sm4key), GetID_IV(), encoding.GetBytes(plaintxt));
            string req = Convert.ToBase64String(enInput);
            Console.WriteLine("加密后req:  " + req);

            // 发送请求
            Hashtable map = new Hashtable();
            map.Add("UID", UID);
            map.Add("ALG", ALG_SM);
            map.Add("DATA", HttpUtility.UrlEncode(req, encoding));
            map.Add("FUNCODE", "DCLISMOD");
            string res = DCHelper.DoPostForm(URL, map);
            Console.WriteLine("res:  " + res);
            try
            {
                Convert.FromBase64String(res);
            }
            catch (Exception e)
            {
                Console.WriteLine("访问返回错误.");
                Console.ReadKey();
                return;
            }

            // 解密请求
            string resplain = encoding.GetString(DCCryptor.CMBSM4DecryptWithCBC(encoding.GetBytes(sm4key), GetID_IV(), Convert.FromBase64String(res)));
            Console.WriteLine("res decrypt: " + resplain);

            // 验签
            JObject object2 = JObject.Parse(resplain);
            JObject object3 = object2["signature"] as JObject;
            string resSign = (string)object3["sigdat"];
            object3["sigdat"] = "__signature_sigdat__";
            object2["signature"] = object3;
            string resSignSource = DCHelper.SerialJsonOrdered(object2);
            Console.WriteLine("验签原文: " + resSignSource);
            Console.WriteLine("验签签名值: " + resSign);
            bool verify = DCCryptor.CMBSM2VerifyWithSM3(GetID_IV(), Convert.FromBase64String(bankpubkey), encoding.GetBytes(resSignSource), Convert.FromBase64String(resSign));
            Console.WriteLine("验签结果: " + verify);
            Console.ReadKey();
        }

        private static byte[] GetID_IV()
        {
            String uid = UID; // 请替换为实际的用户UID
            String userid = uid + "0000000000000000";
            return Encoding.UTF8.GetBytes(userid.Substring(0, 16));
        }
    }
}

File: Demo(Java)-标准/src/main/java/dc/demo/SMDemo.java
package dc.demo;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.HashMap;

import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

/**
 * 国密免前置/SaaS对接示例，本示例仅供参考，不保证各种异常场景运行，请勿直接使用，如有错漏请联系对接人员。运行时，请使用所获取的测试资源替换 用户编号、公私钥、对称密钥、服务商编号等信息。
 */
public class SMDemo {

    private static String URL = "http://cdctest.cmburl.cn:80/cdcserver/api/v2"; // 银行服务地址(测试)
    private static String bankpubkey = "BNsIe9U0x8IeSe4h/dxUzVEz9pie0hDSfMRINRXc7s1UIXfkExnYECF4QqJ2SnHxLv3z/99gsfDQrQ6dzN5lZj0="; // 银行公钥

    private static String privkey = "NBtl7WnuUtA2v5FaebEkU0/Jj1IodLGT6lQqwkzmd2E=";
    private static String sm4key = "VuAzSWQhsoNqzn0K";//"1234567890123456"; // 用户的对称密钥

    private static Base64.Encoder encoder = Base64.getEncoder();
    private static Base64.Decoder decoder = Base64.getDecoder();

    private static final String ALG_SM = "SM"; // 采用国密算法

    private static String UID = "N003261207"; // 测试的用户编号
    
    public static void main(String[] args) throws Exception {
        // 组织发送报文
        JsonObject obj = new JsonObject();
        JsonObject req = new JsonObject();
        JsonObject body = new JsonObject();
        JsonObject head = new JsonObject();
        head.addProperty("funcode", "DCLISMOD");
        head.addProperty("userid", UID);
        head.addProperty("reqid", DCHelper.getTime() + "0000001");
        body.addProperty("buscod", "N02030");
        body.addProperty("TEST", "中文");
        body.addProperty("TEST2", "!@#$%^&*()\\\\///");
        body.addProperty("TEST3", 12345);
        JsonArray array = new JsonArray();
        JsonObject item = new JsonObject();
        item.addProperty("arrItem1", "qaz");
        item.addProperty("arrItem2", 123);
        item.addProperty("arrItem3", true);
        item.addProperty("arrItem4", "中文");

        array.add(item);
        body.add("TEST4", array);
        req.add("head", head);
        req.add("body", body);
        obj.add("request", req);

        // 请求发送接收
        doProcess(obj);
    }

    private static void doProcess(JsonObject jObject) throws Exception {
        JsonObject object = new JsonObject();
        // 签名
        object.addProperty("sigdat", "__signature_sigdat__");
        object.addProperty("sigtim", DCHelper.getTime());
        jObject.add("signature", object);
        String source = DCHelper.serialJsonOrdered(jObject);
        System.out.println("签名原文: " + source);
        byte[] signature1 = DCCryptor.CMBSM2SignWithSM3(getID_IV(), decoder.decode(privkey), source.getBytes(StandardCharsets.UTF_8));
        String sigdat1 = new String(encoder.encode(signature1));
        System.out.println("签名结果: " + sigdat1);
        object.addProperty("sigdat", sigdat1);

        // SM4-CBC加密
        String plaintxt = jObject.toString();
        System.out.println("加密前req:  " + plaintxt);
        byte[] enInput = DCCryptor.CMBSM4EncryptWithCBC(sm4key.getBytes(), getID_IV(), plaintxt.getBytes(StandardCharsets.UTF_8));

        String req = new String(encoder.encode(enInput));
        System.out.println("加密后req:  " + req);

        // 发送请求
        HashMap<String, String> map = new HashMap<>();
        map.put("UID", UID);
        map.put("ALG", ALG_SM);
        map.put("DATA", URLEncoder.encode(req, "utf-8"));
        map.put("FUNCODE", "DCLISMOD");
        String res = DCHelper.doPostForm(URL, map);
        System.out.println("res:  " + res);
        try {
            decoder.decode(res);
        } catch (Exception e) {
            System.err.println("访问返回错误.");
            return;
        }

        // 解密请求
        String resplain = new String(DCCryptor.CMBSM4DecryptWithCBC(sm4key.getBytes(), getID_IV(), decoder.decode(res)), StandardCharsets.UTF_8);
        System.out.println("res decrypt: " + resplain);

        // 验签
        JsonObject object2 = new GsonBuilder().create().fromJson(resplain, JsonObject.class);
        JsonObject object3 = object2.getAsJsonObject("signature");
        String resSign = object3.get("sigdat").getAsString();
        object3.addProperty("sigdat", "__signature_sigdat__");
        object2.add("signature", object3);
        String resSignSource = DCHelper.serialJsonOrdered(object2);
        System.out.println("验签原文: " + resSignSource);
        System.out.println("验签签名值: " + resSign);
        boolean verify = DCCryptor.CMBSM2VerifyWithSM3(getID_IV(), decoder.decode(bankpubkey), resSignSource.getBytes(StandardCharsets.UTF_8), decoder.decode(resSign));
        System.out.println("验签结果: " + verify);
    }

    private static byte[] getID_IV() {
        String uid = UID; // 请替换为实际的用户UID
        String userid = uid + "0000000000000000";
        return userid.substring(0, 16).getBytes();
    }

}
